llama-index-integrations/retrievers/llama-index-retrievers-you/llama_index/retrievers/you/base.py

```diff
@@ -13,24 +15,100 @@
+        # Should deprecate `YOU_API_KEY` in favour of `YDC_API_KEY` for standardization purposes
+        self._api_key = api_key or os.getenv("YOU_API_KEY") or os.environ["YDC_API_KEY"]
         super().__init__(callback_manager)
 
+        if endpoint_type not in ("search", "news"):
+            raise ValueError('`endpoint_type` must be either "search" or "news"')
+
+        # Raise warning if News API-specific fields are set but endpoint_type is not "news"
+        if endpoint_type != "news":
+            news_api_fields = (search_lang, ui_lang, spellcheck)
+            for field in news_api_fields:
+                if field:
+                    warnings.warn(
+                        (
+                            f"News API-specific field '{field}' is set but "
+                            f'`endpoint_type="{endpoint_type}"`. '
+                            "This will have no effect."
+                        ),
+                        UserWarning,
+                    )
+
+        self.endpoint_type = endpoint_type
+        self.num_web_results = num_web_results
+        self.safesearch = safesearch
+        self.country = country
+        self.search_lang = search_lang
+        self.ui_lang = ui_lang
+        self.spellcheck = spellcheck
+
+    def _generate_params(self, query: str) -> Dict[str, Any]:
+        params = {"safesearch": self.safesearch, "country": self.country}
+
+        if self.endpoint_type == "search":
+            params.update(
+                query=query,
+                num_web_results=self.num_web_results,
+            )
+        elif self.endpoint_type == "news":
+            params.update(
+                q=query,
+                count=self.num_web_results,
+                search_lang=self.search_lang,
+                ui_lang=self.ui_lang,
+                spellcheck=self.spellcheck,
+            )
+
+        # Remove `None` values
+        return {k: v for k, v in params.items() if v is not None}
+
     def _retrieve(self, query_bundle: QueryBundle) -> List[NodeWithScore]:
         """Retrieve."""
         headers = {"X-API-Key": self._api_key}
-        results = requests.get(
-            f"https://api.ydc-index.io/search?query={query_bundle.query_str}",
+        params = self._generate_params(query_bundle.query_str)
+        response = requests.get(
+            f"https://api.ydc-index.io/{self.endpoint_type}",
+            params=params,
             headers=headers,
-        ).json()
+        )
+        response.raise_for_status()
+        results = response.json()
 
-        search_hits = ["\n".join(hit["snippets"]) for hit in results["hits"]]
-        return [NodeWithScore(node=TextNode(text=s), score=1.0) for s in search_hits]
+        if self.endpoint_type == "search":
+            retrieved_text = ["\n".join(hit["snippets"]) for hit in results["hits"]]
+        else:  # news endpoint
+            retrieved_text = [
+                article["description"] for article in results["news"]["results"]
```

Comment:
it would be useful to retain the `url` and `age` fields as well so that users can reference the news articles and know how recent the news is. Can you also add those to the retrieved text?

Reply:
Sounds like a good idea. Though I'm thinking of adding it in as `metadata`, instead of as `text` directly. This is to standardize the output with other frameworks which only return `description`, and `metadata` is used by the LLM from the [docs](https://github.com/You-OpenSource/llama_index/blob/43cbd435a9ed6a118a6c05dfbc25c15f96a87665/llama-index-core/llama_index/core/schema.py#L201-L207).

The result would look like this:
```py
>>> results = YouRetriever(endpoint_type='new').retrieve('national parks in the US')
>>> print(results[0].get_content(MetadataMode.ALL)
url: https://brightstandards.com/us-national-parks-list-map/                                                          
age: 5 days ago                                                                                                       
                                                                                                                      
Our 63 US National Parks across America are a collection of some of the most pristine and beautifully preserved landsc
apes in the world. From sea to shining sea, nationally protected treasures are filled with stunning views, remarkable 
wildlife, and historic landmarks. 
```

How does that sound to you, @asna-georgian?

Reply:
Looks sick, i love how clean that is.